crossdock/behavior/trace/behavior.go

```diff
@@ -0,0 +1,254 @@
+type DownstreamCall func(ctx context.Context, target *Downstream) (*Response, error)
+
+// Register function adds JSON and Thrift handlers to the server channel ch
+func (b *Behavior) Register(ch *tchannel.Channel) {
+	b.registerThrift(ch)
+	b.registerJSON(ch)
+}
+
+// Run executes the trace behavior
+func (b *Behavior) Run(t crossdock.T) {
+	logParams(t)
+	sampled, err := strconv.ParseBool(t.Param(sampledParam))
+	if err != nil {
+		t.Fatalf("Malformed param %s: %s", sampledParam, err)
+	}
+	baggage := randomBaggage()
+
+	level1 := &Request{
+		ServerRole: RoleS1,
+	}
+	server1 := t.Param(server1NameParam)
+
+	level2 := &Downstream{
+		ServiceName: t.Param(server2NameParam),
+		ServerRole:  RoleS2,
+		HostPort: fmt.Sprintf("%s:%s",
+			b.serviceToHost(t.Param(server2NameParam)),
+			b.ServerPort,
+		),
+		Encoding: t.Param(server2EncodingParam),
+	}
+	level1.Downstream = level2
+
+	level3 := &Downstream{
+		ServiceName: t.Param(server3NameParam),
+		ServerRole:  RoleS3,
+		HostPort: fmt.Sprintf("%s:%s",
+			b.serviceToHost(t.Param(server3NameParam)),
+			b.ServerPort,
+		),
+		Encoding: t.Param(server3EncodingParam),
+	}
+	level2.Downstream = level3
+
+	resp, err := b.startTrace(t, level1, sampled, baggage)
+	if err != nil {
+		t.Errorf("Failed to startTrace in S1(%s): %s", server1, err.Error())
+		return
+	}
+
+	log.Printf("Response: span=%+v, downstream=%+v", resp.Span, resp.Downstream)
+	traceID := resp.Span.TraceID
+
+	require := crossdock.Require(t)
+	require.NotEmpty(traceID, "Trace ID should not be empty in S1(%s)", server1)
+
+	if validateTrace(t, level1.Downstream, resp, server1, 1, traceID, sampled, baggage) {
+		t.Successf("trace checks out")
+		log.Println("PASS")
+	} else {
+		log.Println("FAIL")
+	}
+}
+
+func logParams(t crossdock.T) {
+	keys := []string{
+		sampledParam,
+		server1NameParam,
+		server2NameParam,
+		server2EncodingParam,
+		server3NameParam,
+		server3EncodingParam,
+	}
+	out := "Execute"
+	for _, key := range keys {
+		out = fmt.Sprintf("%s %s=%s", out, key, t.Param(key))
+	}
+	log.Println(out)
+}
+
+func (b *Behavior) serviceToHost(service string) string {
+	if b.ServiceToHost != nil {
+		return b.ServiceToHost(service)
+	}
+	return service
+}
+
+func (b *Behavior) startTrace(t crossdock.T, req *Request, sampled bool, baggage string) (*Response, error) {
+	span := b.Tracer.StartSpan(req.ServerRole)
+	if sampled {
+		ext.SamplingPriority.Set(span, 1)
+	}
+	span.SetBaggageItem(BaggageKey, baggage)
+	defer span.Finish()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+	defer cancel()
+	ctx = opentracing.ContextWithSpan(ctx, span)
+
+	return b.prepareResponse(t, ctx, req.Downstream)
+}
+
+func validateTrace(
+	t crossdock.T,
+	target *Downstream,
+	resp *Response,
+	service string,
+	level int,
+	traceID string,
+	sampled bool,
+	baggage string,
+) bool {
+	service = fmt.Sprintf("S%d(%s)", level, service)
+	checks := crossdock.Assert(t)
```

Comment:
You called this `checks` but you're using `crossdock.Assert`. Assert and Checks have slightly different behavior: Assert logs even in case of success. This looks like you wanted logging in case of failure only.